### ISSUE TITLE
Unverified: Malformed action parameters after changing associative array parameter format

### ISSUE THREAD

**Issue by johnbillion** [WordPress-Coding-Standards/WordPress-Coding-Standards#1323]
I haven't verified this yet. Might be a `phpcbf` issue, might be a WPCS issue.

The format of the call to the `upgrader_process_complete` action was broken by [core's 42343 changeset](https://core.trac.wordpress.org/changeset/42343).

Before: https://github.com/WordPress/wordpress-develop/blame/ec6a089f986f34a92568ac527f86f33f60fa83c2/src/wp-admin/includes/class-core-upgrader.php#L183

After: https://github.com/WordPress/wordpress-develop/blame/0a63e2ed96f076c42343ba0d55e33217c66456a1/src/wp-admin/includes/class-core-upgrader.php#L192-L197

Placing the `do_action()` parameters on new lines breaks the developer.wordpress.org documentation and results in multiple entries: https://developer.wordpress.org/?s=upgrader_process_complete

**Comment by jrfnl** [comment]
@johnbillion I've just had a look and to be fair, this is not a phpcbf issues, nor a WPCS issue, but an issue with the documentation generator which apparently can't handle multi-line function calls.

**Comment by johnbillion** [comment]
I guess the root issue is that the multi-line array formatting rule affected the whole line, causing the function parameters to appear on new lines, not just the array elements.

**Comment by GaryJones** [comment]
> I guess the root issue is that the multi-line array formatting rule affected the whole line, causing the function parameters to appear on new lines, not just the array elements.

While true, I'd agree that the issue is with the Docs generator.

PSR-2, for instance, covers multi-line function calls:

> Argument lists MAY be split across multiple lines, where each subsequent line
is indented once. When doing so, the first item in the list MUST be on the
next line, and there MUST be only one argument per line.

While:

```php
do_action(
	'upgrader_process_complete', $this, array(
```

satisfies WPCS (as that's what phpcbf produces), so does:

```php
do_action(
	'upgrader_process_complete',
	$this,
	array(
```

and that's what I personally prefer for long function calls (it removes an inconsistency between the two coding standards) - I would expect the documentation generator to be able to handle the different formats.

**Comment by johnbillion** [comment]
I believe the following format is not expected and should be considered a bug:

```php
do_action(
	'upgrader_process_complete', $this, array(
```

It's neither a single-line function call, nor a correctly formatted multi-line function call.

Agreed that the docs parser should support multi-line function calls, but at the same time think we should revert the actions and filters in core to single-liners for consistency and to retain the docs parsing until such time that it gets updated.

**Comment by johnbillion** [comment]
Core ticket :https://core.trac.wordpress.org/ticket/43644

**Comment by jrfnl** [comment]
@johnbillion Just putting it out there: an [RFC](https://github.com/WordPress-Coding-Standards/rfcs) to demand that multi-line function calls have only one argument per line, i.e. "proper" multi-line functions calls, would have my support ;-)

**Comment by azaozz** [comment]
I'd go one step further. Multi-line function and filter calls are harder to read. They should be discouraged, if not completely removed from the WordPress coding standards. If an array needs to be used in them, it should be defined before the call. This makes that bit of code a lot easier to read.

Example:
```
$array = array(
   'position' => $position,
   'checkbox' => $checkbox,
   'mode'     => $mode,
);

$content = apply_filters( 'wp_comment_reply', '', $array );
```

**Comment by jrfnl** [comment]
> Multi-line function and filter calls are harder to read. They should be discouraged, if not completely removed from the WordPress coding standards

I don't agree with that statement. How would you format the following code in that case ?
```php
printf( esc_html__( 'some pretty long sentence with %1$s some %2$s variable replacements', 'text-domain' ), '<a href="' . esc_url( add_query_arg( 'action', 'my_action', admin_url( '/some-page' ) ) . ">, '</a>' );
```

IMHO, the below is far more readable:
```php
printf(
	esc_html__( 'some pretty long sentence with %1$s some %2$s variable replacements', 'text-domain' ),
	'<a href="' . esc_url( add_query_arg( 'action', 'my_action', admin_url( '/some-page' ) ) . ">,
	'</a>'
 );
```

:point_right:  Note, this is just an example, but there are plenty of other code snippets which often lead to ridiculously long lines.

**Comment by azaozz** [comment]
> How would you format the following code in that case?

I don't see any arrays in that bit of code, although if it was me writing it, I'd set 
`esc_url( add_query_arg( 'action', 'my_action', admin_url( '/some-page' ) )` 
to a variable before using it inside the `printf()` call. Makes it even easier to read :)

**Comment by GaryJones** [comment]
Even moving an array into a variable, still means calls like [this](https://github.com/Yoast/wordpress-seo/blob/1712c38b4f423f084ea8ba2c04ab98421e125b10/admin/metabox/class-metabox.php#L204-L207) are going to be longer than is desired. 

The discussion on multi-line function calls isn't just limited to those that contain arrays. But if you want to propose adding a "don't use an array directly in a function call that would make it too long to read / a multi-line function call", then feel free to do the RFC / get it into the handbook and get Core updated to use it :-)

**Comment by azaozz** [comment]
> The discussion on multi-line function calls isn't just limited to those that contain arrays.

See #1330.

> then feel free to do the RFC

Don't think we need an RFC to fix bugs in the rules/parser. The rule about associative arrays being on multiple lines was never intended for inline arrays inside function calls. I was one of the people that added it to the original standards :)

**Comment by GaryJones** [comment]
> The rule about associative arrays being on multiple lines was never intended for inline arrays inside function calls.

Then the rule was incomplete, and introduces an exception (which makes it a less than ideal rule).

Is the issue here then one which requires a sniff to check for direct use of arrays in multiline function calls? Because I can imagine that there is an awful lot of function calls in Core, plugins and themes that already have arrays (and un-necessarily moves us further away from the rest of the PHP community).

**Comment by pento** [comment]
Closing this in favour of #1330.
